File: Travis/indentations.pl
#!/usr/bin/perl
####################################################
#
#          Counts the lines badly indentated.
#
####################################################
use strict;
use warnings;
use English;

if($#ARGV < 0) {
	print "Usage: perl indentations.pl folder1 [folder2...]\n";
	exit(0);
}

my $badFiles = 0;
foreach my $folder (@ARGV) {
	$badFiles += countLinesFolder($folder);
}

exit($badFiles);


###########################################################################
#  Iterates recursively on folders to count the lines badly indentated.
###########################################################################
sub countLinesFolder {
	my $badFiles = 0;
	my $dir = $_[0];

	opendir(DIR, $dir);
	my @entries = readdir(DIR);
	closedir DIR;
	foreach my $entry (@entries) {
		if(-d "$dir/$entry" && $entry ne '..' && $entry ne '.') {
			$badFiles += countLinesFolder("$dir$entry/");
		}
	}

	$badFiles += countLinesFiles($dir);

	return $badFiles;
}

##########################################################
#  Counts the files badly indentated in a folder.
##########################################################
sub countLinesFiles {
	my $dir = $_[0];
	my $badFiles = 0;
	foreach my $fp (glob("$dir*.cpp $dir*.h $dir*.pro")) {
		unless($fp =~ /qcustomplot\.(h|cpp)$/) {
			open my $fh, "<", $fp or die "can't read open '$fp': $OS_ERROR";
			my $nbBadIndentations = 0;
			while(my $line = <$fh>) {
				if($line =~ /^[ ]{3}/) {
					$nbBadIndentations++;
				}
			}
			if($nbBadIndentations > 0) {
				print "$fp contains $nbBadIndentations lines badly indentated.\n";
				$badFiles++;
			}
			close $fh or die "can't read close '$fp': $OS_ERROR";
		}
	}
	return $badFiles;
}

File: Travis/indentations.py
#!/usr/bin/python3

import os
import sys
import re

# Counts the files badly indentated in a folder.
def countLinesFolder(folderPath):
	nbBadFiles = 0;
	# Iterates on all the files.
	for root, subFolders, files in os.walk(folderPath):
		for filename in files:
			if re.search('\.(cpp|h|pro)$', filename) and not re.search('^qcustomplot\.', filename):
				filePath = os.path.join(root, filename)
				nbBadIndentations = countLinesFile(filePath)
				if nbBadIndentations > 0:
					nbBadFiles += 1

	return nbBadFiles


# Counts the lines badly indentated in a file.
def countLinesFile(filePath):
	fp = open(filePath, 'r')
	nbBadIndentations = 0;
	for line in fp:
		if re.search('^[ ]{3}', line):
			nbBadIndentations += 1
	fp.close()
	
	if nbBadIndentations > 0:
		print("%s contains %d lines badly indentated." %(filePath, nbBadIndentations))
	return nbBadIndentations


# Main
nbBadFiles = 0
for folderPath in sys.argv:
	nbBadFiles += countLinesFolder(folderPath)

exit(nbBadFiles)